### PR TITLE
Ensure places/origins are removed when a bookmark with no visits etc is removed.

### PR DESCRIPTION
**Pull request by mhammond**
Fixes a few reports we have on Fenix for this, such as [this one](https://bugzilla.mozilla.org/show_bug.cgi?id=1813898) and I'm sure other dupes exist.

Lina, sorry to do this to you, but you are probably the best reviewer for this :( I'll also ask Ben to help grow more knowledge about places. This is by no means urgent.

### CODE REVIEW

File: components/places/sql/create_shared_triggers.sql [hunk 1]
```diff
@@ -4,13 +4,20 @@
 
 -- This file defines triggers shared between the main and Sync connections.
 
+-- markh doesn't understand why this trigger is needed. It seems to be doing roughly
+-- the same work as moz_updateoriginsinsert_afterdelete_trigger and so we have 2 very
+-- similar triggers for the same insert. A difference between them though is some
+-- frecency stuff - and if you remove this filter, all tests pass except one obscure
+-- one in matcher.rs, where a test site ends up with a frecency of zero rather than the
+-- expected 1999. Both triggers exist in desktop too.
+-- So if anyone can explain this, please update this comment :)
```
[the remark this answers is not in the document]
Comment (linabutler) [reply components/places/sql/create_shared_triggers.sql:13]: @mhammond I figured it out—the [order of the deletes](https://github.com/mozilla/application-services/blob/6c4c65e8ce5808edd12260195e222ed988d780e5/components/places/src/storage/mod.rs#L384-L388) in `delete_pending_temp_tables` is off! 🤯

I think the order of those deletes has to be:

```
DELETE FROM moz_updateoriginsinsert_temp;
DELETE FROM moz_updateoriginsupdate_temp;
DELETE FROM moz_updateoriginsdelete_temp;
```

Specifically, it's important that `DELETE FROM moz_updateoriginsinsert_temp` runs _before_ `DELETE FROM moz_updateoriginsupdate_temp`. Here's what I think is happening, and where the bug is:

When `apply_observation_direct` adds visits for a never-before-visited URL, it does three things:

1. Inserts a row into `moz_places` for the new URL.
2. Inserts rows into `moz_historyvisits` for the new visits.
3. Updates `moz_places.frecency` for the new URL to account for the visits.

(1) fires the `moz_places_afterinsert_trigger_origins` trigger, and (3) fires the `moz_places_afterupdate_frecency_trigger`.

Currently, `delete_pending_temp_tables` does [`DELETE FROM moz_updateoriginsupdate_temp` first](https://github.com/mozilla/application-services/blob/6c4c65e8ce5808edd12260195e222ed988d780e5/components/places/src/storage/mod.rs#L385). That'll fire the `moz_updateoriginsupdate_afterdelete_trigger`, which will try to update `moz_origins.frecency` for the new URL's origin.

_But_, there isn't a row for the URL in `moz_origins` yet—because that row is inserted by the `moz_updateoriginsinsert_afterdelete_trigger`, and that trigger won't fire until we do `DELETE FROM moz_updateoriginsinsert_temp`!

So [this `UPDATE moz_origins` statement](https://github.com/mozilla/application-services/blob/6c4c65e8ce5808edd12260195e222ed988d780e5/components/places/sql/create_shared_triggers.sql#L226-L229) won't do anything, and the initial frecency for the new origin—2000, same as Desktop—will be silently dropped.

When `DELETE FROM moz_updateoriginsinsert_temp` eventually runs, it'll insert a row for the URL's origin (via the `moz_updateoriginsinsert_afterdelete_trigger`), but with an initial origin frecency of 0.

[Desktop's behavior is correct here](https://searchfox.org/mozilla-central/rev/9ac4c7d32fa21ddef5cdf281939228ecfedabf64/toolkit/components/places/History.cpp#991-1009): inserts for new origins are processed before frecency updates for those new origins.

With the order of the deletes fixed, the frecency also changes to match Desktop's! 🎉

TL;DR: I think there are three small fixes:

1. Change the order of the deletes in `delete_pending_temp_tables`.
2. Delete the `moz_places_afterinsert_trigger`, since it's now redundant.
3. Change the expected frecency in the `api::matcher::tests::search` test from 1999 to 2000.

**Review by linabutler — CHANGES_REQUESTED** [review]
Thank you so much for looking at this! 🏆

I wouldn't mind having a fresh look with the origin frecency trigger changes, but your new `moz_cleanup_origin_bookmark_deleted_trigger` and tests look great!

File: components/places/sql/create_shared_triggers.sql [hunk 2]
```diff
@@ -261,6 +268,20 @@ BEGIN
...
+-- This approach may not be suitable for desktop but seems to be for us - see
+-- https://bugzilla.mozilla.org/show_bug.cgi?id=1650511#c41 for more discussion.
+CREATE TEMP TRIGGER moz_cleanup_origin_bookmark_deleted_trigger
+AFTER DELETE ON moz_bookmarks
+BEGIN
+    DELETE FROM moz_places
+        WHERE id = OLD.fk
+        AND foreign_count = 0
+        AND last_visit_date_local = 0
+        AND last_visit_date_remote = 0;
```
Comment (linabutler) [inline components/places/sql/create_shared_triggers.sql:282]: I wonder if we should do the equivalent of `delete_pending_temp_tables` here—or at least a `DELETE FROM moz_updateoriginsdelete_temp`.

I think that would mean the origin would get cleaned up as soon as the bookmark is deleted, and you could avoid the call to `crate::storage::history::delete_everything(&conn)?` in the bookmarks test.

WDYT?

File: components/places/sql/create_shared_triggers.sql (hunk 1 above)
[the remark this answers is not in the document]
Comment (mhammond) [reply components/places/sql/create_shared_triggers.sql]: ah yes, thank you! I actually started with a theory that the order of deletions was wrong, but then convinced myself I was confused about that.

File: components/places/sql/create_shared_triggers.sql (hunk 2 above)
Comment (linabutler) [reply components/places/sql/create_shared_triggers.sql:260]: Oh, rad!

> the trigger on `moz_bookmarks` causes a deletion of `moz_places`, which causes the `moz_places_afterdelete_trigger_origins` trigger.

Yep, you're right!

But the `moz_places_afterdelete_trigger_origins` trigger just inserts a row for the soon-to-be-deleted origin into `moz_updateoriginsdelete_temp`. It doesn't actually delete the origin from `moz_origins`, or update the global frecency stats in `moz_meta`, until you do `DELETE FROM moz_updateoriginsdelete_temp` to fire the `moz_updateoriginsdelete_afterdelete_trigger`.

At first, I thought this would leave the origin behind, because doing `DELETE FROM moz_places` by itself isn't enough to force the `delete_pending_temp_tables` dance. And that's why I thought we needed to force it in the trigger.

But then I realized that `delete_bookmark` calls `delete_pending_temp_tables`! So it's not necessary for your new trigger to do that, because the `delete_pending_temp_tables` call [here](https://github.com/mozilla/application-services/blob/e0b1d5b47b781bb420f6e6fff3bd0ddae408e5cc/components/places/src/storage/bookmarks.rs#L452) will do that work already.

So this is all working correctly now—if in a slightly indirect way that I didn't think all the way through—and you can ignore my suggestion! 😅

**Review by linabutler — APPROVED** [review]
Thanks! My comment was totally a non-issue, sorry to muddle things there 🙈 Land away!
